sql: use leased descriptors for user-defined schemas

thoszhang opened the pull request:
See individual commits for details.

rohany commented on pkg/sql/catalog/descs/collection.go:
```diff
@@ -191,12 +192,17 @@ type Collection struct {
+	// need to resolve the temp schema for the current session, in which case
+	// we can get it from the SessionData. There are some exceptions, like
+	// dropping a database with temp schemas. #53163 is related.
+	// TODO (lucy): Also, while this exists, consider clearing the cache every so
```
The comment that it is never cleared isnt quite true -- we reset this cache when resetting a `descs.Collection`

ajwerner replied:
When does this cache come into play? It seems like it'll happen when resolving a name in a temp schema from a session other than the session which owns it. If that's the case, why do we care about it? Do we even need to cache it? Otherwise, if that's not the case and we use it for also finding temp schemas for the session which owns it, then can't we just not go to the KV and check the session data?

lucy-zhang replied:
Yeah, I originally meant for this to cover both cases (in session and out of session), but I guess you're right that we don't have to cache lookups out of the session, since something like DROP DATABASE CASCADE is a one-off event.

We do also have #53163 as a known issue (the session data only stores the temp schema for the current database, and also never gets updated when the current database changes). So the general cache would bypass this problem. On the other hand, we definitely need to fix that bug soon, at which point this consideration will be irrelevant.

I'll just get rid of the cache. I don't think that will make the temp schema/current database situation much more broken than it is now.

rohany replied:
i agree with andrew that if the cache is only used in drop database cascade then we actually don't need it

lucy-zhang replied:
I removed the cache, RFAL.

rohany commented on pkg/sql/catalog/descs/collection.go:
```diff
@@ -245,6 +251,57 @@ type Collection struct {
+	// reduce the deadline. We use ReadTimestamp() that doesn't return the commit
+	// timestamp, so we need to set a deadline on the transaction to prevent it
+	// from committing beyond the version's expiration time.
+	txn.UpdateDeadlineMaybe(ctx, expiration)
```
this method needs to take in the bool passed from higher levels about whether or not to actually call update

lucy-zhang replied:
Yeah, but in practice that's only needed when we're acquiring descriptors (specifically, types) by ID, and we haven't added that bool argument to any of the methods that resolve by name, which are the exclusive callers of this new method. I don't think it'd be useful to introduce that method argument now, especially since all the callers would pass true as the value. I renamed this function to `getLeasedDescriptorByName` to hopefully make things clearer.

rohany replied:
Thanks for the clarification, that sounds right.

rohany commented on pkg/sql/resolver.go:
```diff
@@ -68,6 +68,34 @@ func (p *planner) ResolveMutableDatabaseDescriptor(
 }
 
+// ResolveUncachedSchemaDescriptor looks up a schema from the store.
+func (p *planner) ResolveUncachedSchemaDescriptor(
```
why do we need this method now?

lucy-zhang replied:
It's just used in a few schema changes where we need an uncached schema descriptor but don't need to update it. I meant for it to be parallel to `ResolveUncachedTableDescriptor`, etc.

rohany commented on pkg/sql/sem/tree/name_resolution.go:
```diff
@@ -576,6 +576,12 @@ type CommonLookupFlags struct {
 
+// SchemaLookupFlags is the flag struct suitable for GetSchema().
+type SchemaLookupFlags struct {
+	CommonLookupFlags
```
I think its time to just add requiremutable to the common lookupflags

lucy-zhang replied:
I was hoping to avoid updating all the existing usages of `RequireMutable` on `tree.ObjectLookupFlags` for now, but I'm willing to reconsider. If I don't do it now, I'll do it this week.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 17 of 18 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-:-MFSFe1Q-DAN3DRkHRIs:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @lucy-zhang)

___
*[pkg/sql/alter_schema.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-MFS5g7q2Xo1kk8riOax:-MFS5g7q2Xo1kk8riOay:bzi9wav) ([raw file](https://github.com/cockroachdb/cockroach/blob/b279272c2bcb11ecdd1f77b608e5e71181f8fb14/pkg/sql/alter_schema.go#L55)):*
> ```Go
> 		return nil, pgerror.Newf(pgcode.InvalidSchemaName, "cannot modify schema %q", n.Schema)
> 	case catalog.SchemaUserDefined:
> 		// TODO (rohany): Check permissions here.
> ```

Should we create an issue to track this?
___
*[pkg/sql/catalog/descs/collection.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-MFSCS0k9L1GqSUS6uiX:-MFSBTFX6IeZuS_ANsvK:byij878) ([raw file](https://github.com/cockroachdb/cockroach/blob/b279272c2bcb11ecdd1f77b608e5e71181f8fb14/pkg/sql/catalog/descs/collection.go#L260)):*
> ```Go
> 
> // getLeasedDescriptor return a leased descriptor valid for the transaction,
> // acquiring one if necessary. Due to a bug in lease acquisition for dropped
> ```

are we tracking this bug in an issue?
___
*[pkg/sql/catalog/descs/collection.go, line 445 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-MFSCVJO9qwoYBva_nNX:-MFSCVJO9qwoYBva_nNY:bn73rnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b279272c2bcb11ecdd1f77b608e5e71181f8fb14/pkg/sql/catalog/descs/collection.go#L445)):*
> ```Go
> 	ctx context.Context, txn *kv.Txn, dbID descpb.ID, schemaName string, flags tree.SchemaLookupFlags,
> ) (*schemadesc.MutableSchemaDescriptor, error) {
> 	if log.V(2) {
> ```

nit: `log.ExpensiveLoggingEnabled(ctx, 2)` or `VEventf` so we get this in traces
___
*[pkg/sql/catalog/descs/collection.go, line 506 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-MFSEBzl0LWVbDXUxVMs:-MFSEBzl0LWVbDXUxVMt:b90ytuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b279272c2bcb11ecdd1f77b608e5e71181f8fb14/pkg/sql/catalog/descs/collection.go#L506)):*
> ```Go
> 	}
> 
> 	// TODO (lucy): Replace this with a lookup to the parent database's Schema
> ```

Should we track this with an issue?


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53247)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @rohany)

___
*[pkg/sql/alter_schema.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-MFS5g7q2Xo1kk8riOax:-MFSW2Qu3uTKFSH3qFn_:b-ifpetz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b279272c2bcb11ecdd1f77b608e5e71181f8fb14/pkg/sql/alter_schema.go#L55)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Should we create an issue to track this?
</blockquote></details>

Probably. I want to take stock of all the remaining privileges-related issues soon.
___
*[pkg/sql/catalog/descs/collection.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-MFSCS0k9L1GqSUS6uiX:-MFSQz7m0BNf_n7V2l6L:b-juvcek) ([raw file](https://github.com/cockroachdb/cockroach/blob/b279272c2bcb11ecdd1f77b608e5e71181f8fb14/pkg/sql/catalog/descs/collection.go#L260)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

are we tracking this bug in an issue?
</blockquote></details>

We are now: #53299
___
*[pkg/sql/catalog/descs/collection.go, line 445 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-MFSCVJO9qwoYBva_nNX:-MFSW-Za4O-OVEsJL-RU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b279272c2bcb11ecdd1f77b608e5e71181f8fb14/pkg/sql/catalog/descs/collection.go#L445)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: `log.ExpensiveLoggingEnabled(ctx, 2)` or `VEventf` so we get this in traces
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/collection.go, line 506 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-MFSEBzl0LWVbDXUxVMs:-MFSVe-OBIylir4imqAd:bg2v0ah) ([raw file](https://github.com/cockroachdb/cockroach/blob/b279272c2bcb11ecdd1f77b608e5e71181f8fb14/pkg/sql/catalog/descs/collection.go#L506)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Should we track this with an issue?
</blockquote></details>

I filed #53301, which hopefully we can close soon.


<!-- Sent from Reviewable.io -->

rohany commented on pkg/sql/create_schema.go:
```diff
@@ -97,6 +97,7 @@ func (p *planner) createUserDefinedSchema(params runParams, n *tree.CreateSchema
 		Name:       n.Schema,
 		ID:         id,
 		Privileges: privs,
+		Version:    1,
```
Should all of our `NewMutableCreated...Descriptor` should set the version to 1? We've (me) missed this in alot of places

lucy-zhang commented:
I'm still looking at the `TestClusterQueriesTxnData` failure, btw. There's what appears to be an extra lease acquisition showing up in the KV trace that I don't understand.

ajwerner reviewed (commented):
> I'm still looking at the TestClusterQueriesTxnData failure, btw. There's what appears to be an extra lease acquisition showing up in the KV trace that I don't understand.

Interesting, curious to hear what that's about. 

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> still

> Reviewed 4 of 18 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-:-MFTIAkw50Q2TMB_Z0hC:b-d9n4zf)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @rohany)

___
*[pkg/sql/create_schema.go, line 100 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/53247#-MFRIGNK_z-gaxL2swno-r4-100:-MFTHsUQ2j4SL4IABwhm:bri3v6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c05e13f74f0499f7812abaf256732d33589e5a2/pkg/sql/create_schema.go#L100)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Should all of our `NewMutableCreated...Descriptor` should set the version to 1? We've (me) missed this in alot of places
</blockquote></details>

The one place to be careful is I think in databases we want to keep it zero until the cluster version is upgrade. That whole thing generally needs some love. 


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
So, `TestClusterQueriesTxnData` is failing for pretty much expected reasons: We now introduce an extra round of lease acquisition to every `dbname.tablename` lookup. Before this change, a trace for `select * from t.t` would look like this:

```
             timestamp             |       age       |                         message                          |                            tag                             |                location                 |    operation     | span
-----------------------------------+-----------------+----------------------------------------------------------+------------------------------------------------------------+-----------------------------------------+------------------+-------
  2020-08-24 01:54:28.243633+00:00 | 00:00:11.184227 | querying next range at /NamespaceTable/30/1/50/0/"t"/4/1 | [n1,client=127.0.0.1:51046,hostssl,user=root,txn=34147a6a] | kv/kvclient/kvcoord/range_iter.go:159   | dist sender send |    3
  2020-08-24 01:54:28.243691+00:00 | 00:00:11.184285 | r26: sending batch 1 Get to (n1,s1):1                    | [n1,client=127.0.0.1:51046,hostssl,user=root,txn=34147a6a] | kv/kvclient/kvcoord/dist_sender.go:1811 | dist sender send |    3
  2020-08-24 01:54:28.244092+00:00 | 00:00:11.184686 | querying next range at /Table/60/1                       | [n1,client=127.0.0.1:51046,hostssl,user=root]              | kv/kvclient/kvcoord/range_iter.go:159   | exec stmt        |    1
  2020-08-24 01:54:28.244226+00:00 | 00:00:11.18482  | Scan /Table/60/{1-2}                                     | [n1,client=127.0.0.1:51046,hostssl,user=root]              | sql/row/kv_batch_fetcher.go:315         | table reader     |    7
  2020-08-24 01:54:28.244246+00:00 | 00:00:11.18484  | querying next range at /Table/60/1                       | [n1,client=127.0.0.1:51046,hostssl,user=root,txn=34147a6a] | kv/kvclient/kvcoord/range_iter.go:159   | dist sender send |    9
  2020-08-24 01:54:28.244272+00:00 | 00:00:11.184866 | r66: sending batch 1 Scan to (n1,s1):1                   | [n1,client=127.0.0.1:51046,hostssl,user=root,txn=34147a6a] | kv/kvclient/kvcoord/dist_sender.go:1811 | dist sender send |    9
  2020-08-24 01:54:28.244447+00:00 | 00:00:11.185041 | fetched: /t/primary/583806657514602497/x -> /1           | [n1,client=127.0.0.1:51046,hostssl,user=root]              | sql/row/fetcher.go:1209                 | table reader     |    7
  2020-08-24 01:54:28.24455+00:00  | 00:00:11.185144 | rows affected: 1                                         | [n1,client=127.0.0.1:51046,hostssl,user=root]              | sql/conn_executor_exec.go:605           | exec stmt        |    1
```

Now it looks like this (the extra transaction is causing the test to fail, but that's not really relevant to the problem at hand):

```
             timestamp             |       age       |                         message                          |                            tag                             |                location                 |    operation     | span
-----------------------------------+-----------------+----------------------------------------------------------+------------------------------------------------------------+-----------------------------------------+------------------+-------
  2020-08-24 01:46:39.479714+00:00 | 00:00:11.638138 | querying next range at /NamespaceTable/30/1/50/0/"t"/4/1 | [n1,client=127.0.0.1:50971,hostssl,user=root,txn=5dbfcb62] | kv/kvclient/kvcoord/range_iter.go:159   | dist sender send |    3
  2020-08-24 01:46:39.479766+00:00 | 00:00:11.63819  | r26: sending batch 1 Get to (n1,s1):1                    | [n1,client=127.0.0.1:50971,hostssl,user=root,txn=5dbfcb62] | kv/kvclient/kvcoord/dist_sender.go:1811 | dist sender send |    3
  2020-08-24 01:46:39.480009+00:00 | 00:00:11.638433 | querying next range at /Table/2/1/50/"t"/3/1             | [n1,client=127.0.0.1:50971,hostssl,user=root,txn=5dbfcb62] | kv/kvclient/kvcoord/range_iter.go:159   | dist sender send |    6
  2020-08-24 01:46:39.480038+00:00 | 00:00:11.638462 | r6: sending batch 1 Get to (n1,s1):1                     | [n1,client=127.0.0.1:50971,hostssl,user=root,txn=5dbfcb62] | kv/kvclient/kvcoord/dist_sender.go:1811 | dist sender send |    6
  2020-08-24 01:46:39.480207+00:00 | 00:00:11.638631 | querying next range at /NamespaceTable/30/1/50/0/"t"/4/1 | [n1,client=127.0.0.1:50971,hostssl,user=root,txn=00ad65ba] | kv/kvclient/kvcoord/range_iter.go:159   | dist sender send |    9
  2020-08-24 01:46:39.480231+00:00 | 00:00:11.638655 | r26: sending batch 1 Get to (n1,s1):1                    | [n1,client=127.0.0.1:50971,hostssl,user=root,txn=00ad65ba] | kv/kvclient/kvcoord/dist_sender.go:1811 | dist sender send |    9
  2020-08-24 01:46:39.48051+00:00  | 00:00:11.638934 | querying next range at /Table/60/1                       | [n1,client=127.0.0.1:50971,hostssl,user=root]              | kv/kvclient/kvcoord/range_iter.go:159   | exec stmt        |    1
  2020-08-24 01:46:39.480619+00:00 | 00:00:11.639043 | Scan /Table/60/{1-2}                                     | [n1,client=127.0.0.1:50971,hostssl,user=root]              | sql/row/kv_batch_fetcher.go:315         | table reader     |   13
  2020-08-24 01:46:39.480637+00:00 | 00:00:11.639061 | querying next range at /Table/60/1                       | [n1,client=127.0.0.1:50971,hostssl,user=root,txn=00ad65ba] | kv/kvclient/kvcoord/range_iter.go:159   | dist sender send |   15
  2020-08-24 01:46:39.480662+00:00 | 00:00:11.639086 | r66: sending batch 1 Scan to (n1,s1):1                   | [n1,client=127.0.0.1:50971,hostssl,user=root,txn=00ad65ba] | kv/kvclient/kvcoord/dist_sender.go:1811 | dist sender send |   15
  2020-08-24 01:46:39.480818+00:00 | 00:00:11.639242 | fetched: /t/primary/583805094527664129/x -> /1           | [n1,client=127.0.0.1:50971,hostssl,user=root]              | sql/row/fetcher.go:1209                 | table reader     |   13
  2020-08-24 01:46:39.480897+00:00 | 00:00:11.639321 | rows affected: 1                                         | [n1,client=127.0.0.1:50971,hostssl,user=root]              | sql/conn_executor_exec.go:605           | exec stmt        |    1
```
What I didn't originally get is that after attempting and failing to get a lease on a nonexistent descriptor, we also subsequently try to look up the descriptor from the store in the original transaction, which would explain the second `querying next range at /NamespaceTable/30/1/50/0/"t"/4/1`. I'm not sure why we do this. It seems like we should just return an error, but this has been the behavior of the lease manager for some time.

Anyway, the point of all this is that we're doing even more reads than I thought, in the absence of caching negative results for these schema lookups.

lucy-zhang commented:
I added an explicit public schema to the table names in the test, which works around this for now.

I'll merge this now because it's closer to the desired final state of things and there's no overall benefit in holding off for beta 1, even though it makes name resolution performance worse than the status quo (which is already bad). #53290 should fix it, but I don't know if it'll make it into beta 1. I'll keep working on it.

TFTRs

bors r+

lucy-zhang commented:
bors r+

lucy-zhang commented:
Rebasing on top of #53278....

bors r+

lucy-zhang commented:
bors r-

Commit:
sql: use leased descriptors for user-defined schemas

This commit introduces the use of leased descriptors for user-defined
schemas. The API in sql/resolver.go has been extended to support mutable
schema descriptor access for schema changes, and name resolution on
leased object descriptors now uses leased schemas.

Note that the preexisting `desc.Collection.schemaCache` is now
restricted to temp schemas. It can't be removed entirely because we
currently need to resolve temp schemas not bound to the current session
while dropping databases (see inline comments).

Release note: None

Commit:
sql/catalog: extract a getLeasedDescriptor method in descs.Collection

Minor refactor to pull out repeated shared logic.

Release note: None

lucy-zhang commented:
bors r+
